File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
from collections import Counter


def get_parent(p, x):
    if p[x] != x:
        p[x] = get_parent(p, p[x])
    return p[x]


def union_parent(p, a, b):
    a = get_parent(p, a)
    b = get_parent(p, b)
    if a < b:
        p[b] = a
    elif a > b:
        p[a] = b
    else:
        return


def solution(n, computers):
    parent = [i for i in range(n + 1)]
    for i, a in enumerate(computers, start=1):
        for j, b in enumerate(a, start=1):
            if b:
                union_parent(parent, i, j)
    for i, e in enumerate(parent):
        parent[i] = get_parent(parent, i)
    return len(Counter(parent[1:]))

File: 프로그래머스/lv3/43162. 네트워크/네트워크.java
import java.util.*;

class Solution {
    int[] parent;
    
    public int solution(int n, int[][] computers) {
        parent = new int[n];
        for (int i = 0; i < n; i++)
            parent[i] = i;
        
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++){
                if (computers[i][j] == 1)
                    unionParent(i, j);
            }
        }
        
                    
        Set<Integer> set = new HashSet<>();
        for (int i = 0; i < n; i++){
            set.add(getParent(i));
        }
        return set.size();
    }
    
    int getParent(int x){
        if (parent[x] != x)
            parent[x] = getParent(parent[x]);
        return parent[x];
    }
    
    void unionParent(int a, int b){
        a = getParent(a);
        b = getParent(b);
        if (a > b)
            parent[a] = b;
        else if (a < b)
            parent[b] = a;
        else
            return;
    }
}